backend/arm64/emit.mlp

```diff
@@ -1268,6 +1268,125 @@ let data l =
+  module Asm_targets.Asm_directives.Make(struct
+    let emit_line = emit_line
+
+    let get_file_num file_name =
+      Emitaux.get_file_num ~file_emitter file_name
+
+    let debugging_comments_in_asm_files =
+      !Flambda_backend_flags.dasm_comments
+
+    module D = struct
+      type constant =
+        | Int64 of Int64.t
+        | Label of string
+        | Add of constant * constant
+        | Sub of constant * constant
+
+      let rec string_of_constant const =
+        match const with
+        | Int64 n -> Int64.to_string n
+        | Label s -> s
+        | Add (c1, c2) ->
+          Printf.sprintf "(%s + %s)"
+            (string_of_constant c1) (string_of_constant c2)
+        | Sub (c1, c2) ->
+          Printf.sprintf "(%s - %s)"
+            (string_of_constant c1) (string_of_constant c2)
+
+      let const_int64 num = Int64 num
+      let const_label str = Label str
+      let const_add c1 c2 = Add (c1, c2)
+      let const_sub c1 c2 = Sub (c1, c2)
+
+      type data_type =
+        | NONE
+        | DWORD
+        | QWORD
+        | VEC128
+
+      let file = file_emitter
+
+      let loc ~file_num ~line ~col ?discriminator () =
+        ignore discriminator;
+        emit_line (Printf.sprintf ".loc %d %d %d" file_num line col)
+
+      let comment str =
+        emit_line (Printf.sprintf "; %s" str)
+
+      let label ?data_type str =
+        let _ = data_type in
+        emit_line (Printf.sprintf "%s:" str)
+
+      let section ?delayed:_ name flags args =
+        match name, flags, args with
+        | [".data" ], _, _ -> emit_line "\t.data"
+        | [".text" ], _, _ -> emit_line "\t.text"
+        | name, flags, args ->
+          emit_string (Printf.sprintf "\t.section %s"
+            (String.concat "," name));
+          begin match flags with
+          | None -> ()
+          | Some flags -> emit_string (Printf.sprintf ",%S" flags)
+          end;
+          begin match args with
+          | [] -> ()
+          | _ ->
+            emit_string (Printf.sprintf ",%s" (String.concat "," args))
+          end;
+          emit_string "\n"
+
+      let text () = emit_line "\t.text"
+
+      let new_line () = emit_line ""
+
+      let global sym = emit_line (Printf.sprintf "\t.globl %s" sym)
+
+      let protected _sym = ()  (* CR mshinwell: fixme *)
+
+      let type_ sym typ_ = emit_line (Printf.sprintf "\t.type %s,%s" sym typ_)
+
+      let byte const =
+        emit_line
+          (Printf.sprintf "\t.byte %s" (string_of_constant const))
+
+      let word const =
+        emit_line
+          (Printf.sprintf "\t.short %s" (string_of_constant const))
+
+      let long const =
+        emit_line
+          (Printf.sprintf "\t.long %s" (string_of_constant const))
+
+      let qword const =
+        emit_line
+          (Printf.sprintf "\t.quad %s" (string_of_constant const))
+
+      let bytes str =
+        emit_line (Printf.sprintf "\t.ascii %S" str)
```

Comment:
Is there a risk the `str` could be
really large? (I vaguely remember
an assembler whose behaviour
was quite bad -quadratic?)

Reply:
Not sure if there is a problem here, but this matches what the emitter code does further down the file, so I think it's ok.

Reply:
(I vaguely remember we had to split
long string to reduce the time it takes
for the assembly to produce a `o` file.)